sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -113,18 +196,75 @@ RT::PiKernel ProgramManager::getOrCreateKernel(OSModuleHandle M,
     std::cerr << ">>> ProgramManager::getOrCreateKernel(" << M << ", "
               << getRawSyclObjImpl(Context) << ", " << KernelName << ")\n";
   }
-  RT::PiProgram Program = getBuiltOpenCLProgram(M, Context);
+
+  auto Program = getBuiltOpenCLProgram(M, Context);
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
-  std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> &CachedKernels =
-      Ctx->getCachedKernels();
-  std::map<string_class, RT::PiKernel> &KernelsCache = CachedKernels[Program];
-  RT::PiKernel &Kernel = KernelsCache[KernelName];
-  if (!Kernel) {
-    PI_CALL(piKernelCreate)(Program, KernelName.c_str(), &Kernel);
+
+  using PiKernelT = kernel_program_cache::PiKernelT;
+  using KernelWithBuildState = kernel_program_cache::KernelWithBuildStateT;
+
+  kernel_program_cache &Cache = *Ctx->getKernelProgramCache().get();
+  bool InsertionTookPlace;
+  KernelWithBuildState *KernelWithState;
```

Comment:
```suggestion
  KernelWithBuildState *KernelWithState = nullptr;
```

Reply:
It's initialized pretty soon after allocation. Don't see any if this change is obligatory